User/v mhoffman/sources panel intro (#1175)
MicrosoftDocs/edge-developer: microsoft-edge/devtools-guide-chromium/sources/index.md

:::image type="complex" source="../media/sources-filesystem-tab.msft.png" alt-text="The Filesystem tab, for a Workspace" lightbox="../media/sources-filesystem-tab.msft.png":::
   The **Filesystem** tab, for a Workspace
:::image-end:::  

By default, when you edit a file in the **Sources** tool, those changes are lost when you refresh the page.  The **Sources** tool works with a copy of the front-end resources that are returned by the web server.  When you modify these front-end files that are returned by the server, the changes don't persist, because you didn't change the source files.  You need to also apply your edits in your actual source code, and then re-deploy to the server.

By using a Workspace, changes that you make to your front-end code are preserved across page refreshes.  When you edit the front-end code that's returned by the server, the Sources tool also applies your edits to your local source code.  For other users to see your changes, you redeploy your changed source files to the server.

> Workspaces work well when the JavaScript code that's returned by the server is the same as your local JavaScript source code.  Workspaces don't work as well when your workflow involves transformations on your JavaScript source code, such as minification or compilation.

Reviewer: ```suggestion
Workspaces work well when the JavaScript code that's returned by the server is the same as your local JavaScript source code.  Workspaces don't work as well when your workflow involves transformations on your source code, such as minification or [TypeScript](https://www.typescriptlang.org/) compilation.
```

   See the License for the specific language governing permissions and
   limitations under the License.  -->

# Sources tool overview  

Use the **Sources** tool to view, modify, and debug JavaScript that's returned by the server, and to inspect the resources that make up the current webpage.  To use the **Sources** tool as a development environment, add source files to a Workspace.  The Sources tool has three panes:

| Pane | Actions |
|---|---|

> | **Explorer** pane | Navigate among the resources that are returned from the server to construct the current webpage.  Select files, images, and other resources, and view their paths.  Optionally, set up a local Workspace to save changes directly to source files. |

Author: Globally changed Explorer pane to Navigator pane.

#### Sources tool without a Workspace

In contrast to using a Workspace, when you use the Sources tool in the default way, you edit a transient copy of the front-end code that's returned by the web server.  This non-Workspace approach supports fast iterations when you are prototyping JavaScript code and don't want to go through the build process every time you want to view the rendered result of a code edit.

By default, when you edit a file in the **Sources** tool, those changes are lost when you refresh the page.  In the default approach, the **Sources** tool works with a copy of the client-side resources that are returned by the web server.  When you modify these client-side files that are returned by the server, the changes don't persist, because you didn't change the source files.

When you find a fix that works for the client-side JavaScript that is returned by the server, you then need to apply that fix, or an equivalent of that fix, in your actual source code and then re-deploy to the server.

#### Mapping minified code to your source code, to work with readable source code

> <!-- From a user point of view, is Workspace related to mapping minified code to source code? -->

Author: @zoherghadyali    From a user point of view, is Workspace related to mapping minified code to source code?

If your client-side source code is transformed, so that the server returns different, transformed, hard-to-read (minified) front-end code to the browser, in some cases, you can set up mapping from your source code to your compiled code, so that when you view, debug, and edit the transformed client-side JavaScript code that's returned from the server, the code looks familiar and readable.

Source Maps from preprocessors cause DevTools to load your original JavaScript source files in addition to your minified, transformed JavaScript files that are returned by the server.  You then use your original source files to set breakpoints and step through code.  Meanwhile, Microsoft Edge is actually running your minified code. 

The mapping approach keeps your client-side code readable and debuggable even after you combine, minify, or compile it.
For more information, navigate to [Map preprocessed code to source code][DevToolsJavaScriptSourceMaps].

#### Complex transformations from source code to compiled front-end code

> <!-- todo: Does this section have to do with Workspace?  if not, maybe move this section elsewhere, to preserve the series of 5 sections for the 5 tabs of the Explorer pane. -->

Author: @zoherghadyali - Does this section (Complex transformations from source code to compiled front-end code) have to do with Workspace? if not, maybe move this section elsewhere

These three approaches are equivalent:

*   Temporarily adding the statements `console.log(sum)` and `console.log(typeof sum)` in the code, where `sum` is in-scope.
*   Issuing the statements `sum` and `console.log(typeof sum)` in the **Console** pane of the DevTools, when the debugger is paused where `sum` is in-scope.
*   Setting the **Watch** expressions `sum` and `typeof sum` in the **Debugger** pane.

When the variable `sum` is in-scope, `sum` and its value are automatically shown in the **Scope** section of the **Debugger** pane, and are also overlaid in the Editor pane where `sum` is calculated.  So you probably wouldn't need to define a Watch expression for `sum`.

The debugger gives a richer, more flexible display and environment than a `console.log` statement.  For example, in the debugger, as you step through the code, you may display and change the values of all currently defined properties and variables.  You can also issue JavaScript statements in the **Console**, such as to change values in an array that's in-scope.  (To display the Console, select **Esc**.)

Breakpoints and Watch expressions are preserved when you refresh the page.

> ### Using the debugger of Visual Studio Code

Reviewer: ```suggestion
### Debug from Visual Studio Code directly
```

| CSS | View and edit. |
| HTML | View and edit. |
| Images | View. |

The following subsections are related to the Editor pane:
*   [Reformat a minified JavaScript file with pretty-print](#reformat-a-minified-javascript-file-with-pretty-print)
*   [Apply or save changes](#apply-or-save-changes)
*   [Source mapping](#source-mapping)
*   [Editor commands on the Command Menu](#editor-commands-on-the-command-menu)
*   [Edit an HTML file](#edit-an-html-file)
*   [Edit a CSS file](#edit-a-css-file)

### Reformat a minified JavaScript file with pretty-print

To use pretty-print to reformat a file to make it readable, click the **Pretty print** button \(![Format](../media/format-icon.msft.png)\) at the bottom of the Editor pane.  For more information, navigate to the section
[Reformat a minified JavaScript file with pretty-print][DevToolsJavaScriptReferenceReformat] within the article "Use the debugger features".

### Apply or save changes

For JavaScript changes to take effect, select `Control`+`S` \(Windows, Linux\) or `Command`+`S` \(macOS\).  DevTools does not re-run a script, so the only JavaScript changes that take effect are those that you make inside of functions.  

> For example, in the following figure, notice how `console.log('A')` does not run, whereas `console.log('B')` does.  If DevTools re-runs the entire script after making the change, <!-- why/when would DevTools re-run the entire script?  above says "DevTools does not re-run a script" --> then the text `A` is logged to the **Console**.  

Author: @zoherghadyali - why/when would DevTools re-run the entire script? above says "DevTools does not re-run a script"

   Editing JavaScript in the **Editor** pane  
:::image-end:::

The **Editor** pane has the following level of support for various file types:

| File Type | Supported Actions |
|---------|---------|
| JavaScript | View, edit, and debug. |
| CSS | View and edit. |
| HTML | View and edit. |
| Images | View. |

The following subsections are related to the Editor pane:
*   [Reformat a minified JavaScript file with pretty-print](#reformat-a-minified-javascript-file-with-pretty-print)
*   [Mapping minified code to your source code, to work with readable source code](#mapping-minified-code-to-your-source-code-to-work-with-readable-source-code)
*   [Apply or save changes](#apply-or-save-changes)
*   [Editor commands on the Command Menu](#editor-commands-on-the-command-menu)
*   [Edit an HTML file](#edit-an-html-file)
*   [Edit a CSS file](#edit-a-css-file)

### Reformat a minified JavaScript file with pretty-print

To use pretty-print to reformat a file to make it readable, click the **Pretty print** button \(![Format](../media/format-icon.msft.png)\) at the bottom of the Editor pane.  For more information, navigate to the section

> [Reformat a minified JavaScript file with pretty-print][DevToolsJavaScriptReferenceReformat] within the article "Use the debugger features".

Author: No; the sentence contains a link to that article.  The "Reformat" link at the start of the sentence is a link to a subsection within the "Use the debugger features" article.  
Details about the link are at the bottom of the present file:
[DevToolsJavaScriptReferenceReformat]: ../javascript/reference.md#reformat-a-minified-javascript-file-with-pretty-print "Reformat a minified JavaScript file with pretty-print - Use the debugger features | Microsoft Docs"

For example, in the following figure, notice how `console.log('A')` does not run, whereas `console.log('B')` does.  If DevTools re-runs the entire script after making the change, <!-- why/when would DevTools re-run the entire script?  above says "DevTools does not re-run a script" --> then the text `A` is logged to the **Console**.  

By default, your edits are discarded when you refresh the page.  To learn how to save the changes to your file system, navigate to [Edit files with Workspaces][DevtoolsGuideChromiumWorkspacesIndex].

### Editor commands on the Command Menu

To go to a line number or symbol in the file: 

In the **Explorer** pane, select the ellipses (...) (**More options**), select **Open File** (**Ctrl+P**), then type any of the following characters:

*   `?` to view options.
*   \: to **Go to line**.
*   \@ to **Go to symbol**.

For more information, navigate to [Run commands with the Microsoft Edge DevTools Command Menu](../command-menu/index.md).

> ### Edit an HTML file

Reviewer: Also I'd like to move Edit CSS up above Edit HTML. There's not really a good reason someone would edit HTML in the Sources tool. In order of usefulness, it's probably:

1. Edit JavaScript
2. Edit CSS
3. Edit HTML

*   In the **Elements** tool, [Edit CSS font styles and settings in the Styles pane][DevToolsInspectStylesEditFonts] (recommended in most cases).
*   In the **Sources** tool, in the **Editor** pane, edit the CSS file as a file listing that directly shows the file.

The Sources tool supports directly editing a CSS file.  For example, if you edit the CSS file from the tutorial [Edit files with Workspaces][DevtoolsGuideChromiumWorkspacesIndex] to match the style rule below, the `H1` element in the upper left of the rendered webpage changes to green:

```css
h1 {
  color: green;
}  
```

:::image type="complex" source="../media/edit-css.msft.png" alt-text="Edit CSS in the Editor pane to change the text color of the H1 heading to green" lightbox="../media/edit-css.msft.png":::
   Edit CSS in the **Editor** pane to change the text color of the `H1` heading to green  
:::image-end:::  

CSS changes take effect immediately; in most cases, no saving is needed.

**Related articles:**

*   [Edit CSS font styles and settings in the Styles pane][DevToolsInspectStylesEditFonts]

*   [DevTools for beginners: Get started with CSS][DevToolsBeginnersCss] - tutorial

> ### Opening the DevTools code editor when using a tool other than the Sources tool

Reviewer: ```suggestion
### Display sources even when using a different tool
```

> The main place to use the code editor of DevTools is within the **Sources** tool.  But sometimes you need to access other tools, such as the **Elements** tool or the **Console** panel, while making changes to your site.  There is a way to have the DevTools code editor open alongside other panels.  

Reviewer: ```suggestion
The main place to view source files in the DevTools is within the **Sources** tool.  But sometimes you need to access other tools, such as **Elements** or **Console**, while viewing or editing your source files.  Use the **Quick Sources** tool in the [Drawer](https://docs.microsoft.com/microsoft-edge/devtools-guide-chromium/customize/#drawer).
```

1.  Choose a tool other than the **Sources** tool, such as the **Elements** tool.  
1.  Select `Control`+`Shift`+`P` \(Windows, Linux\) or `Command`+`Shift`+`P` \(macOS\).  The **Command Menu** opens.  

> 1.  Type `QS`, then choose **Show Quick Source**.  At the bottom of the DevTools window, the **Quick Source** panel appears, displaying the last file you edited in the **Sources** tool, within the code editor from the **Sources** tool.  

Reviewer: I think for consistency, I'd prefer if our instructions were "Type `Quick Source`" and then have them select and run the command from there. It's a little more robust this way if we ever add a command in the future that also has a Q and an S in it

1.  Select `Control`+`P` \(Windows, Linux\) or `Command`+`P` \(macOS\) to open the **Open File** dialog.  

## The Debugger pane, to debug JavaScript code

Use the JavaScript Debugger to step through the JavaScript code that's returned by the server (or stored locally in a Workspace). 
The Debugger includes the **Debugger** pane, along with breakpoints that are set on lines of code in the **Editor** pane.

<!-- "sources-paused-breakpoint.msft.png" is not currently directly used. It's the upstream source for a couple images, including the below image. -->

:::image type="complex" source="../media/sources-paused-breakpoint-highlight-debug-pane.msft.png" alt-text="The Debugger pane of the Sources tool  " lightbox="../media/sources-paused-breakpoint-highlight-debug-pane.msft.png":::
   The **Debugger** pane of the Sources tool  
:::image-end:::  

> Step through experimental changes to the code, while watching any JavaScript expressions you specify.  Watch and manually change variable values, and automatically show which variables are currently in-scope for the current line of code.

Reviewer: ```suggestion
Step through the code, while watching any JavaScript expressions you specify.  Watch and manually change variable values, and automatically show which variables are currently in-scope for the current line of code.
```

The Debugger supports standard debugging actions, such as:  
*   Setting breakpoints, to pause code.
*   Stepping through code.
*   Viewing and editing properties and variables.
*   Watching the values of JavaScript expressions.
*   Viewing the call stack (the sequence of function calls so far).

> The JavaScript Debugger in DevTools, in the **Debugger** pane, is designed to look and feel and work like 

Reviewer: ```suggestion
The JavaScript Debugger in DevTools, in the **Debugger** pane, is designed to look, feel, and work like 
```

The following subsections cover the **Debugger** pane or breakpoints:
*   [The basic approach to using a debugger](#the-basic-approach-to-using-a-debugger)
*   [Advantages of the debugger's Watch and Scope over console.log](#advantages-of-the-debuggers-watch-and-scope-over-consolelog)
*   [Articles about debugging](#articles-about-debugging)

### The basic approach to using a debugger

To troubleshoot where your JavaScript is going wrong, you can insert `console.log()` statements in the **Editor** pane.  Another approach, that's more powerful (yet can also be simple), is to use the debugger of Microsoft Edge DevTools.

The basic approach to using a debugger is to set a breakpoint (an intentional stopping place in the code), run the code at full speed up to the line of code on which the breakpoint is set, and then step through the lines of code, starting at that breakpoint.  

For example, the tutorial [Get started with debugging JavaScript][DevtoolsGuideChromiumJavascriptIndex] uses an existing, simple webpage 
[Demo: Get Started Debugging JavaScript with Microsoft Edge (Chromium) DevTools][DevtoolsGlitchMeDebugJsGetStarted], that contains a few form controls.  

> As shown in the tutorial, to get into the "Paused" mode of the debugger, do something like the following steps:

Reviewer: Isn't this content duplicative with the one we have in the demo in the debugging doc you linked? I think it either makes more sense to direct customers to the debugging doc or to move the content from that doc here. I'm not sure copying it over again, in less depth, adds value

Author: I streamlined this section and repositioned it with more emphasis on the general case than on the tutorial demo in particular.

The difference between the info in this article and in the demo tutorial article is the zoom level: overview (the least you need to know) vs. detail.  The "overview" pattern is used throughout this Overview article: this article presents a compact (1-screen) summary of a given feature of the Sources tool, and then links to a detailed (multi-page) article in case the reader wants more information.

The debugger gives a richer, more flexible display and environment than a `console.log` statement.  For example, in the Debugger, as you step through the code, you may display and change the values of all currently defined properties and variables, and issue JavaScript statements in the **Console**.  (To display the Console, select **Esc**.)

Breakpoints and Watch expressions are preserved when you refresh the page.

### Articles about debugging

The following articles cover the Debugger pane and breakpoints:

*   [Get started with debugging JavaScript in Microsoft Edge DevTools][DevtoolsGuideChromiumJavascriptIndex] - A tutorial (with screen captures), using an existing, simple project.

*   [Use the debugger features][DevToolsJavaScriptReference] - How to use the Debugger to set breakpoints, step through code, view and modify variable values, watch JavaScript expressions, and view the call stack (the sequence of function calls so far).

*   [Pause your code with breakpoints][DevToolsJavaScriptBreakpoints] - How to set basic and specialized breakpoints in the debugger.

> ## Extensions for Visual Studio Code, as an alternative to DevTools Sources tool

Reviewer: I think we can cut this whole section. I don't think it adds much and we have our own docs for the VS Code extensions anyways
